### PR TITLE
Confidentialize project and organization names in reports

### PR DESCRIPTION
**Pull request by gigxz**
* Enforce confidentiality rules in non-HUD reports.
* Fix bugs where project names and organizations were being confidentialized even when the user is permitted to see them. [see comment](https://github.com/greenriver/hmis-warehouse/pull/2267#issuecomment-1160233687)

Report categories completed:
- [x] Audit
- [x] CAS
- [x] Census
- [x] Data Quality
- [x] Exports
- [x] Health
- [x] Operational
- [x] Performance
- [x] Population Dashboards
- [x] Public

### CODE REVIEW

File: app/controllers/warehouse_reports/client_details/last_permanent_zips_controller.rb [hunk 1]
```diff
@@ -28,8 +28,9 @@ def index
       end
 
       @clients = population_service_history_source.
-        joins(:client, :enrollment, :project).
-        includes(:client, :enrollment, :project).
+        joins(:client, :enrollment, :project, :organization).
+        includes(:client, :enrollment, :project, :organization).
+        preload(:organization). # FIXME why isnt this working, still a bunch of organization queries
```
Comment (gigxz) [inline app/controllers/warehouse_reports/client_details/last_permanent_zips_controller.rb:33]: Trying to get rid of N+1 that's happening [here](https://github.com/greenriver/hmis-warehouse/blob/21aeb8ddaf1fc6d7ff778767d886a417b5eda5b3/app/views/warehouse_reports/client_details/last_permanent_zips/index.haml#L47), but this doesn't seem to fix it. @eanders can we look at this together at some point?

File: app/controllers/warehouse_reports/entry_exit_service_controller.rb [hunk 2]
```diff
@@ -42,10 +55,12 @@ def columns
         FirstName: c_t[:FirstName],
         LastName: c_t[:LastName],
         destination_id: wc_t[:destination_id],
-        ProjectID: e_t[:ProjectID],
-        ProjectName: p_t[:ProjectName],
+        project_id: p_t[:id],
```
Comment (gigxz) [inline app/controllers/warehouse_reports/entry_exit_service_controller.rb:58]: unrelated bug fix, we were incorrectly passing a `ProjectID` instead of `id` into `project_path`, so all the links were broken

**Comment by gigxz** [comment]
https://github.com/greenriver/hmis-warehouse/pull/2267/commits/c0f3d13c67ce14a9c6b9f3741c6cc0903013e9b9 fixes an issue where Project and Organization names were always being confidentialized on the User Access and Group Access assignment pages. I updated them so that Org and Project names are _never_ confidentialized there. This bug was a result of the fact that `organization.name` and `project.name` now default to confidentializing–– so we are sometimes confidentializing things where we shouldn't be. I'm trying to catch all instances of this, but I'm sure some will slip through the cracks (like this one, almost https://github.com/greenriver/hmis-warehouse/pull/2267/commits/8be3fd85da258879a68167968126f47f76e89507).


The linked commit makes a change to `ViewableEntities`, so it impacts several controllers, all of which require `can_edit_users`. It also affects the groups controller which requires `can_edit_access_groups`. So, is it OK to show confidential names to users with those admin permissions? If not, I can update it to confidentialize project/orgs based on the user's confidentiality perm. What do you think @eanders?

**Comment by eanders** [comment]
@gigxz, I think this makes sense.  If I'm understanding the code, this would always expose the name of organizations and projects within the context of assigning access to them and when listing assignments.  As long as this doesn't expose the names on reports to people who can edit users/assignments, which I don't see happening, we're good.

**Comment by gigxz** [comment]
It does expose names of assigned Projects/Organizations on the User Permission Report (under "view details"). That's [this code](https://github.com/greenriver/hmis-warehouse/pull/2267/commits/8be3fd85da258879a68167968126f47f76e89507).  It doesn't look like we require any special permissions to view that report.

**Comment by eanders** [comment]
That report is fine.  It just lists the viewable entities.

File: app/controllers/warehouse_reports/missing_values_controller.rb [hunk 3]
```diff
@@ -216,10 +216,13 @@ def all_sources
...
           ot = organizations.arel_table
           pt = projects.arel_table
-          sql = data_sources.importable.joins(organizations: :projects).
-            merge(projects.viewable_by(user)).
-            select(st[:id], st[:name], st[:short_name], ot[:id], ot[:OrganizationName], pt[:id], pt[:ProjectName]).
-            to_sql
+
+          scope = data_sources.importable.joins(organizations: :projects).merge(projects.viewable_by(user))
+          scope = scope.merge(organizations.non_confidential) unless user.can_view_confidential_enrollment_details?
+          scope = scope.merge(projects.non_confidential) unless user.can_view_confidential_enrollment_details?
```
Comment (eanders) [inline app/controllers/warehouse_reports/missing_values_controller.rb]: This is really minor, but maybe:
```suggestion
          scope = data_sources.importable.joins(organizations: :projects).merge(projects.viewable_by(user))
          unless user.can_view_confidential_enrollment_details?
            scope = scope.merge(organizations.non_confidential).
              merge(projects.non_confidential) 
          end
```

File: app/controllers/warehouse_reports/recidivism_controller.rb [hunk 4]
```diff
@@ -8,24 +8,20 @@ module WarehouseReports
   class RecidivismController < ApplicationController
     include WarehouseReportAuthorization
     include ArelHelper
+
     def index
-      columns = [:client_id, :project_name, :first_date_in_program, :last_date_in_program, :move_in_date, :computed_project_type, :id]
       @filter = ::Filters::DateRange.new(date_range_options)
-      ph_scope = ph_source.open_between(start_date: @filter.start, end_date: @filter.end).distinct
+      ph_scope = ph_source.joins(:project, project: [:organization]).open_between(start_date: @filter.start, end_date: @filter.end).distinct
```
Comment (eanders) [inline app/controllers/warehouse_reports/recidivism_controller.rb]: I _think_ this is the same result
```suggestion
      ph_scope = ph_source.joins(project: :organization).open_between(start_date: @filter.start, end_date: @filter.end).distinct
```

File: app/controllers/warehouse_reports/recidivism_controller.rb [hunk 5]
```diff
@@ -8,24 +8,20 @@ module WarehouseReports
...
       @ph_clients = ph_scope.pluck(*columns).
-        map do |row|
-          Hash[columns.zip(row)]
-        end.
+        map { |row| to_hash_confidentialized(row) }.
         group_by { |row| row[:client_id] }
 
-      @homeless_clients = homeless_source.
+      @homeless_clients = homeless_source.joins(:project, project: [:organization]).
```
Comment (eanders) [inline app/controllers/warehouse_reports/recidivism_controller.rb]: Here too, if that's true.

File: app/models/concerns/arel_helper.rb [hunk 6]
```diff
@@ -538,6 +548,20 @@ def cast(exp, as)
       nf 'CAST', [exp.as(as)]
     end
 
+    def confidentialized_project_name(column)
```
Comment (eanders) [inline app/models/concerns/arel_helper.rb:552]:
```suggestion
    # NOTE: you must join project and organization for this to work.
    def confidentialized_project_name(column)
```

File: app/controllers/warehouse_reports/recidivism_controller.rb (hunk 4 above)
Comment (gigxz) [reply app/controllers/warehouse_reports/recidivism_controller.rb]: ah yeah, that generates the same SQL. I did this in a bunch of places, I'll update them

File: app/models/grda_warehouse/warehouse_reports/time_homeless_for_exit.rb [hunk 7]
```diff
@@ -73,13 +73,14 @@ def average_days
           )
         end
         # Layer on the first enrollment in PH if it is earlier
-        clients_with_housed_enrollments.order(e_t[:MoveInDate].asc).pluck(:id, e_t[:MoveInDate], p_t[:ProjectName]).each do |client_id, date, project_name|
+        clients_with_housed_enrollments.order(e_t[:MoveInDate].asc).pluck(:id, e_t[:MoveInDate], p_t[:ProjectName], p_t[:confidential], o_t[:confidential]).each do |client_id, date, project_name, p_confidential, o_confidential|
```
Comment (eanders) [inline app/models/grda_warehouse/warehouse_reports/time_homeless_for_exit.rb]: any reason not to use `bool_or(p_t[:confidential], o_t[:confidential]).as('confidential')` here?

Comment (gigxz) [reply app/models/grda_warehouse/warehouse_reports/time_homeless_for_exit.rb]: no, just that I did this before I invented `bool_or`. I'll update.

File: app/views/clients/_enrollment_table.haml [hunk 8]
```diff
@@ -58,7 +58,8 @@
         %td
           %span{data: @client.program_tooltip_data_for_enrollment(e)}
             - if include_links
-              = link_to e[:project_name], project_path(id: e[:project_id])
+              - can_view_project_page = e[:project_name] != GrdaWarehouse::Hud::Project.confidential_project_name || current_user.can_edit_projects?
```
Comment (eanders) [inline app/views/clients/_enrollment_table.haml]: Is include links something like `can_view_projects?`  I _think_ `can_view_project_page` should be something like:
```suggestion
              - can_view_project_page = current_user.can_edit_projects?
              - can_view_project_page ||=  e[:project_name] != GrdaWarehouse::Hud::Project.confidential_project_name && current_user.can_view_projects? 
```

File: app/views/dashboards/base/_index.haml [hunk 9]
```diff
@@ -2,10 +2,7 @@
 = render 'index_title'
 - unless GrdaWarehouse::Config.get(:multi_coc_installation)
   = render 'housed_overall'
-- if @html
```
Comment (eanders) [inline app/views/dashboards/base/_index.haml:5]: Pretty sure this was to prevent the picker from showing up in PDFs.  (It's totally possible this is handled later, but wanted to check.)

File: app/views/data_quality_reports/version_four/project/support.haml [hunk 10]
```diff
@@ -13,11 +13,14 @@
     %thead
       %tr
         - @data[:headers].each do |t|
+          - next if t == 'Project ID'
```
Comment (eanders) [inline app/views/data_quality_reports/version_four/project/support.haml:16]: Why are we hiding Project IDs?

File: app/views/warehouse_reports/hmis_cross_walks/index.xlsx.axlsx [hunk 11]
```diff
@@ -17,9 +17,9 @@ wb.add_worksheet(name: 'Projects') do |sheet|
       [
```
Comment (eanders) [inline app/views/warehouse_reports/hmis_cross_walks/index.xlsx.axlsx]: This report should probably ignore the confidentiality as well.  It doesn't include any client detail, and the entire point is to be able to lookup IDs from the LSA results and figure out which project or inventory record is causing trouble in the sending HMIS.

File: app/views/warehouse_reports/initiatives/show.haml [hunk 12]
```diff
@@ -16,9 +16,9 @@
       Chosen Options
 
 .tab-content
-  %div.tab-pane.active.show.fade{role: :tabpanel, id: 'settings', aria:{labelledby:"settings"}}
+  %div.tab-pane.fade{role: :tabpanel, id: 'settings', aria:{labelledby:"settings"}}
     = render 'settings'
-  %div.tab-pane.fade{role: :tabpanel, id: 'outcomes', aria:{labelledby:"outcomes"}}
+  %div.tab-pane.active.show.fade{role: :tabpanel, id: 'outcomes', aria:{labelledby:"outcomes"}}
```
Comment (eanders) [inline app/views/warehouse_reports/initiatives/show.haml:21]: Thanks.  I was thinking the same.

File: app/views/dashboards/base/_index.haml (hunk 9 above)
Comment (gigxz) [reply app/views/dashboards/base/_index.haml:5]: This template is only every invoked if `@html` is true, so it was dead code:
https://github.com/greenriver/hmis-warehouse/blob/aaf97a25704742593a51a0134cd1b99e18b04646/app/views/dashboards/base/index.haml#L1-L4

File: app/views/data_quality_reports/version_four/project/support.haml (hunk 10 above)
Comment (gigxz) [reply app/views/data_quality_reports/version_four/project/support.haml:16]: Project IDs weren't visible in the report before, I just had to add them to the plucked fields so I could check if it's confidential (by checking against cached list of confidential IDs). It's kind of hacky, but using an `includes(:project)` ended up being way more complicated (at least with my rails skill level :)

If this general approach sounds okay, we could also just show the Project IDs in the table, or drop them from the controller instead of the view.

File: app/views/clients/_enrollment_table.haml (hunk 8 above)
Comment (gigxz) [reply app/views/clients/_enrollment_table.haml]: Ok. `include_links` sometimes gets passed as `true` (from enrollments_history) and other times gets passed as  `can_view_projects?`, I'm not sure why.

File: app/views/data_quality_reports/version_four/project/support.haml (hunk 10 above)
Comment (eanders) [reply app/views/data_quality_reports/version_four/project/support.haml:16]: This is fine (as would be including them in the page.)  Just curious since it looked a bit odd.

File: app/views/clients/_enrollment_table.haml (hunk 8 above)
Comment (gigxz) [reply app/views/clients/_enrollment_table.haml]: Your logic makes sense to me, and is easier to read, updated in https://github.com/greenriver/hmis-warehouse/pull/2267/commits/67df096dd4a687a20ffbb697e9e866140edab08e

File: app/views/warehouse_reports/hmis_cross_walks/index.xlsx.axlsx (hunk 11 above)
Comment (gigxz) [reply app/views/warehouse_reports/hmis_cross_walks/index.xlsx.axlsx]: OK, updated with https://github.com/greenriver/hmis-warehouse/pull/2267/commits/fc2f2181351dd31a68e75a43c506e854e6a670a7 and added to list of exceptions here https://github.com/greenriver/hmis-warehouse/pull/2267/commits/2a440ab62d1f688ae6efce46a0bb970f39897d59

File: app/jobs/warehouse_reports/run_active_veterans_job.rb [hunk 13]
```diff
@@ -15,7 +16,7 @@ def perform(params)
 
       user = User.find(params[:current_user_id])
       report.user_id = user.id
-      report.parameters[:visible_projects] = if user.can_view_all_reports?
+      report.parameters[:visible_projects] = if GrdaWarehouse::DataSource.can_see_all_data_sources?(user)
```
Comment (gigxz) [inline app/jobs/warehouse_reports/run_active_veterans_job.rb:19]: I think this must have been a mistake to check `can_view_all_reports?` here, becuase we use this later on to render "Unlimited" vs "Limited," and the "Limited" value shows a tooltip saying something like "this report is limited to the following projects that are viewable to the user: x,y,z"
@eanders

Comment (eanders) [reply app/jobs/warehouse_reports/run_active_veterans_job.rb:19]: You are correct, that was a historical use of the permission.  Not sure `can_see_all_data_sources?` is correct, but it is more correct.  It should probably just always use `GrdaWarehouse::Hud::Project.viewable_by(user).pluck(:id, :ProjectName)`.  At some level, there will need to be an awareness that very few people can see all projects and your reports will be limited to the projects you can see.

File: app/models/grda_warehouse/hud/client.rb [hunk 14]
```diff
@@ -1783,11 +1783,12 @@ def last_seen_in_type(type, include_confidential_names: false)
       return nil unless type.in?(GrdaWarehouse::Hud::Project::RESIDENTIAL_PROJECT_TYPES.keys + [:homeless])
 
       service_history_enrollments.ongoing.
-        joins(:service_history_services, :project).
+        joins(:service_history_services, :project, :organization).
```
Comment (eanders) [inline app/models/grda_warehouse/hud/client.rb:1786]: should this be?
```suggestion
        joins(:service_history_services, project: :organization).
```
It only matters if we reference `enrollment.project.organization` vs `enrollment.organization`

Comment (eanders) [reply app/models/grda_warehouse/hud/client.rb:1786]: never mind, just here so we can reference `o_t[:confidential]`

File: app/jobs/warehouse_reports/run_active_veterans_job.rb (hunk 13 above)
Comment (gigxz) [reply app/jobs/warehouse_reports/run_active_veterans_job.rb:19]: Ok. At this point this is ignored anyway, since we removed the `Limited?` column from the view.
